include/Refine.h

```diff
@@ -125,444 +95,258 @@ template<typename real_t, int dim> class Refine
-
-                        send_cnt[i] = send_additional[i].size();
-                        for(typename std::set< DirectedEdge<index_t> >::const_iterator it=send_additional[i].begin(); it!=send_additional[i].end(); ++it) {
-                            _mesh->send[i].push_back(it->id);
-                            _mesh->send_halo.insert(it->id);
-                        }
-                    }
-
-                    // Additional code for centroidal vertices.
-                    if(dim==3)
-                    {
-                        for(int i=0; i<nprocs; ++i) {
-                            recv_cnt[i] += cidRecv_additional[i].size();
-                            for(typename std::set<Wedge>::const_iterator it=cidRecv_additional[i].begin(); it!=cidRecv_additional[i].end(); ++it) {
-                                _mesh->recv[i].push_back(it->cid);
-                                _mesh->recv_halo.insert(it->cid);
-                            }
-
-                            send_cnt[i] += cidSend_additional[i].size();
-                            for(typename std::set<Wedge>::const_iterator it=cidSend_additional[i].begin(); it!=cidSend_additional[i].end(); ++it) {
-                                _mesh->send[i].push_back(it->cid);
-                                _mesh->send_halo.insert(it->cid);
-                            }
-                        }
-                    }
-
-                    // Update global numbering
-                    _mesh->update_gappy_global_numbering(recv_cnt, send_cnt);
-
-                    // Now that the global numbering has been updated, update send_map and recv_map.
-                    for(int i=0; i<nprocs; ++i)
-                    {
-                        for(typename std::set< DirectedEdge<index_t> >::const_iterator it=recv_additional[i].begin(); it!=recv_additional[i].end(); ++it)
-                            _mesh->recv_map[i][_mesh->lnn2gnn[it->id]] = it->id;
-
-                        for(typename std::set< DirectedEdge<index_t> >::const_iterator it=send_additional[i].begin(); it!=send_additional[i].end(); ++it)
-                            _mesh->send_map[i][_mesh->lnn2gnn[it->id]] = it->id;
-
-                        // Additional code for centroidals.
-                        if(dim==3) {
-                            for(typename std::set<Wedge>::const_iterator it=cidRecv_additional[i].begin(); it!=cidRecv_additional[i].end(); ++it)
-                                _mesh->recv_map[i][_mesh->lnn2gnn[it->cid]] = it->cid;
-
-                            for(typename std::set<Wedge>::const_iterator it=cidSend_additional[i].begin(); it!=cidSend_additional[i].end(); ++it)
-                                _mesh->send_map[i][_mesh->lnn2gnn[it->cid]] = it->cid;
-
-                            cidRecv_additional[i].clear();
-                            cidSend_additional[i].clear();
-                        }
+                }
+                if (lengths[iEdg] < 0.9*max_length_cavity) {
```

Comment:
This is what seems to make it work, and yet kills the purpose of the quality based algorithm: only refine an edge if it's (almost) the longest of the cavity (irrespective of if the other edges were tagged or not)